File: app/src/main/java/com/norvera/confession/data/AppRepository.java
package com.norvera.confession.data;

import com.norvera.confession.data.models.CommandmentEntry;
import com.norvera.confession.data.models.ExaminationActiveEntry;
import com.norvera.confession.data.models.ExaminationEntry;
import com.norvera.confession.data.models.GuideEntry;
import com.norvera.confession.data.models.InspirationEntry;
import com.norvera.confession.data.models.PrayersEntry;
import com.norvera.confession.utils.AppExecutor;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.sqlite.db.SimpleSQLiteQuery;


public class AppRepository {
    private final AppExecutor executor;
    private final AppDatabase appDatabase;

    public AppRepository(AppDatabase appDatabase, AppExecutor executor) {
        this.executor = executor;
        this.appDatabase = appDatabase;
    }

    public LiveData<List<CommandmentEntry>> loadCommandmentsForReligious() {
        return appDatabase.commandmentDao().loadCommandmentsForReligious();
    }

    public LiveData<List<CommandmentEntry>> loadCommandmentsForLayPerson() {
        return appDatabase.commandmentDao().loadCommandmentsForLayPerson();
    }

    public LiveData<CommandmentEntry> loadCommandmentById(long id) {
        return appDatabase.commandmentDao().selectCommandmentById(id);
    }

    public void addCommandment(CommandmentEntry commandmentEntry) {
        executor.diskIO().execute(() -> appDatabase.commandmentDao().addCommandment(commandmentEntry));
    }


    public LiveData<List<ExaminationEntry>> loadAllExaminationsForCommandmentAndUser(SimpleSQLiteQuery query) {
        return appDatabase.examinationDao().loadAllExaminationsForCommandmentAndUser(query);
    }

    public void updateCountForEntry(ExaminationEntry examinationEntry) {
        executor.diskIO().execute(() -> appDatabase.examinationDao().updateEntry(examinationEntry));
    }

    public LiveData<List<GuideEntry>> allGuidesForId(int guideId) {
        return appDatabase.guideDao().allGuidesForId(guideId);
    }

    public LiveData<List<PrayersEntry>> loadAllPrayers() {
        return appDatabase.prayersDao().loadAllPrayers();
    }

    public LiveData<List<ExaminationActiveEntry>> loadAllExaminationsWithCount() {
        return appDatabase.examinationDao().loadAllExaminationsWithCount();
    }

    public LiveData<InspirationEntry> getInspirationForId(int id) {
        return appDatabase.inspirationDao().getInspirationForId(id);
    }

    public void decrementCountForEntry(ExaminationEntry examinationEntry) {
        executor.diskIO().execute(() -> appDatabase.examinationDao().decrementCount(examinationEntry));
    }
}

File: app/src/main/java/com/norvera/confession/data/AppRepository.kt
package com.norvera.confession.data

import androidx.lifecycle.LiveData
import androidx.sqlite.db.SimpleSQLiteQuery
import com.norvera.confession.data.models.*
import com.norvera.confession.utils.AppExecutor
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.withContext


class AppRepository(private val appDatabase: AppDatabase, private val executor: AppExecutor) {

    fun loadCommandmentsForReligious(): LiveData<List<CommandmentEntry>> {
        return appDatabase.commandmentDao().loadCommandmentsForReligious()
    }

    fun loadCommandmentsForLayPerson(): LiveData<List<CommandmentEntry>> {
        return appDatabase.commandmentDao().loadCommandmentsForLayPerson()
    }

    fun loadCommandmentById(id: Long): LiveData<CommandmentEntry> {
        return appDatabase.commandmentDao().selectCommandmentById(id)
    }

    fun addCommandment(commandmentEntry: CommandmentEntry) {
        executor.diskIO().execute { appDatabase.commandmentDao().addCommandment(commandmentEntry) }
    }


    fun loadAllExaminationsForCommandmentAndUser(query: SimpleSQLiteQuery): LiveData<List<ExaminationEntry>> {
        return appDatabase.examinationDao().loadAllExaminationsForCommandmentAndUser(query)
    }

    suspend fun updateCountForEntry(examinationEntry: ExaminationEntry) {
        withContext(IO) {
            appDatabase.examinationDao().updateEntry(examinationEntry)
        }
    }

    fun allGuidesForId(guideId: Int): LiveData<List<GuideEntry>> {
        return appDatabase.guideDao().allGuidesForId(guideId)
    }

    fun loadAllPrayers(): LiveData<List<PrayersEntry>> {
        return appDatabase.prayersDao().loadAllPrayers()
    }

    fun loadAllExaminationsWithCount(): LiveData<List<ExaminationActiveEntry>> {
        return appDatabase.examinationDao().loadAllExaminationsWithCount()
    }

    fun getInspirationForId(id: Int): LiveData<InspirationEntry> {
        return appDatabase.inspirationDao().getInspirationForId(id)
    }

    suspend fun decrementCountForEntry(examinationEntry: ExaminationEntry) {
        withContext(IO) {
            appDatabase.examinationDao().decrementCount(examinationEntry)
        }
    }
}
